### PR TITLE
dev: add Sentry integration

### PR DESCRIPTION
**Pull request by digitalcora**
I tested this in dev by deploying versions of the scripts that intentionally threw errors (the helper for creating "named" errors and the cutoff for processor errors are a result of this). Everything seems to show up correctly now, including stack traces and errors being tagged with the specific Lambda that generated them (packager/processor).

I didn't set up source maps yet, since this has some extra complexity (it requires setting up "releases") and I'm not sure whether it's worth our while. It's straightforward to bundle the scripts locally and get the same combined `.js` file that is shipped to Lambda, and thus appears in Sentry stack traces. The only disadvantage might be that Sentry doesn't know which lines are "our code" and which are libraries, for grouping purposes.

### CODE REVIEW

File: src/processor.ts [hunk 1]
```diff
@@ -25,10 +28,22 @@ const transformRecord = ({ recordId, data }: EventRecord): ResultRecord => {
...
     };
   } catch (error) {
+    maybeCaptureError(error);
     return { recordId, result: "ProcessingFailed", data };
   }
 };
 
+// Only capture an error once per run; if many records throw errors and we try
+// to capture them all, the capturing itself can take a long time and cause the
+// Lambda to time out
```
Comment (bklebe) [inline src/processor.ts:38]: question: will an exception getting thrown cause the lambda to restart? Seems like we don't want to lose a lot of errors, because it could be a bug in the lambda and not just a problem with the record.

File: src/errors.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Convenience function for constructing errors where the `name` property has a
+ * given value. This enables easier identification and grouping in e.g. Sentry.
+ */
+export const exception = (name: string, message: string) => {
+  const error = new Error(message);
+  error.name = name;
```
Comment (bklebe) [inline src/errors.ts:7]: suggestion: instead of using a helper function, would it be possible to change the `name` property in the `Error` subclasses you created before?

File: src/packager.ts [hunk 3]
```diff
@@ -10,54 +10,57 @@ import {
...
+    const env = struct({ ...process.env }, Environment);
+    const client = makeS3Client(env);
+    const {
+      S3_BUCKET: bucket,
+      S3_PREFIX_OUTPUT: outputPrefix,
+      S3_PREFIX_SOURCE: sourcePrefix,
+    } = env;
+
+    const serviceDay = localFromISO(time).minus({ days: 1 }).toISODate();
+    const outputLabel = serviceDay.replace(/-/g, "");
```
Comment (bklebe) [inline src/packager.ts:37]: question: does it make sense to separate the reformatting out into a separate commit here? Not for review purposes, but for history?

File: src/errors.ts (hunk 2 above)
Comment (digitalcora) [reply src/errors.ts:7]: As far as I can tell, there's no straightforward way to do this other than typing out something like:

```js
class MyError extends Error {
  function constructor(...args) {
    super(...args);
    this.name = "MyError";
    Error.captureStackTrace(this, this.constructor);
  }
}
```

...for every error class you want to have. Given I'm not sure this approach has any advantages over the current one (do you know of any?), I'd prefer to avoid the noise and duplication.

Comment (bklebe) [reply src/errors.ts:7]: Hmmm, I was hoping it could be done somehow without having to write a constructor. This feels like a weird pattern to me coming from other object-oriented languages with exceptions, but if it's the simplest way to achieve this with Sentry then it makes sense. Is the specific Sentry issue documented anywhere? I tried looking at `ocs_saver` in our Sentry account but didn't see any errors.

File: src/processor.ts (hunk 1 above)
Comment (digitalcora) [reply src/processor.ts:38]: Since we're catching all exceptions within the processing of each record and instead returning `ProcessingFailed` for that record, the Lambda will not stop or restart when this happens. The point of this once-per-run code is to intentionally "lose" all such errors other than the first, because otherwise the Lambda times out and and we don't see anything at all in Sentry (unknown why we don't at least get a few errors; possibly the separate error-sending process never gets a chance to run, as JS is single-threaded).

Assuming we could report more than one error per run, I'm not sure what the value of this would be. Could you elaborate on what information you're concerned we might lose?

File: src/errors.ts (hunk 2 above)
Comment (digitalcora) [reply src/errors.ts:7]: I don't know if it's explicitly documented that Sentry uses the `name` property of errors to title and group them. I might not call it an "issue" exactly — from what I've seen, other libraries and even Node itself use this property to identify errors.

File: src/processor.ts (hunk 1 above)
Comment (digitalcora) [reply src/processor.ts:38]: > Well, if processing different records fails for different reasons we might want to know what they are.

We should be able to determine when this happens by looking in the `failed` prefix in the S3 bucket, as every record we assign a status of `ProcessingFailed` will be dumped in there. So Sentry can alert us to the fact that there's at least one issue, and then we can look at S3 to see all the issues (or, all the records that caused issues, but we should be able to reproduce the issues from there).

> Is this just a limitation of the Firehose model?

We could increase the timeout for the Lambda — the default is 3 seconds, but Firehose supports a timeout of up to 5 minutes. If you think it's worth trying, with some additional Terraform work, we could do so.

File: src/packager.ts (hunk 3 above)
Comment (digitalcora) [reply src/packager.ts:37]: If by "reformatting" you mean the whole re-indentation of this function... I'm not sure? When the `wrapHandler` call is present, Prettier formats it like this; when it isn't, Prettier formats it the other way. I suppose we could throw in a temporary `// prettier-ignore`, then remove it in the subsequent commit. There is always the whitespace-ignoring diff mode, though.

Comment (bklebe) [reply src/packager.ts:37]: Oh, ok. I didn't realize that the indentation level actually increased.

File: src/processor.ts (hunk 1 above)
Comment (bklebe) [reply src/processor.ts:38]: > So Sentry can alert us to the fact that there's at least one issue, and then we can look at S3 to see all the issues (or, all the records that caused issues, but we should be able to reproduce the issues from there).

Yeah, if we're able to run them through this to reproduce, that should be enough.

**Review by bklebe — APPROVED** [review]
